libraries/AC_Autorotation/AC_Autorotation.cpp

```diff
@@ -2,375 +2,524 @@
+    }
+    // TODO: need to improve the handling of the velocity NED not ok case
+    return vel_NED.x;
 }
 
-
-// update speed controller
-void AC_Autorotation::update_forward_speed_controller(void)
+#if HAL_LOGGING_ENABLED
+// Logging of lower rate autorotation specific variables. This is meant for stuff that
+// doesn't need a high rate, e.g. controller variables that are need for tuning.
+void AC_Autorotation::log_write_autorotation(void) const
 {
-    // Specify forward velocity component and determine delta velocity with respect to time
-    _speed_forward = calc_speed_forward(); //(cm/s)
-
-    _delta_speed_fwd = _speed_forward - _speed_forward_last; //(cm/s)
-    _speed_forward_last = _speed_forward; //(cm/s)
-
-    // Limiting the target velocity based on the max acceleration limit
-    if (_cmd_vel < _vel_target) {
-        _cmd_vel += _accel_max * _dt;
-        if (_cmd_vel > _vel_target) {
-            _cmd_vel = _vel_target;
-        }
-    } else {
-        _cmd_vel -= _accel_max * _dt;
-        if (_cmd_vel < _vel_target) {
-            _cmd_vel = _vel_target;
-        }
+    // enum class for bitmask documentation in logging
+    enum class AC_Autorotation_Landed_Reason : uint8_t {
+        LOW_SPEED = 1<<0, // true if below 1 m/s
+        LAND_COL  = 1<<1, // true if collective below land col min
+        IS_STILL  = 1<<2, // passes inertial nav is_still() check
+    };
+
+    uint8_t reason = 0;
+    if (_landed_reason.min_speed) {
+        reason |= uint8_t(AC_Autorotation_Landed_Reason::LOW_SPEED);
+    }
+    if (_landed_reason.land_col) {
+        reason |= uint8_t(AC_Autorotation_Landed_Reason::LAND_COL);
+    }
+    if (_landed_reason.is_still) {
+        reason |= uint8_t(AC_Autorotation_Landed_Reason::IS_STILL);
     }
 
-    // get p
-    _vel_p = _p_fw_vel.get_p(_cmd_vel-_speed_forward);
+    // @LoggerMessage: AROT
+    // @Vehicles: Copter
+    // @Description: Helicopter AutoROTation (AROT) information
+    // @Field: TimeUS: Time since system startup
+    // @Field: LR: Landed Reason state flags
+    // @FieldBitmaskEnum: LR: AC_Autorotation_Landed_Reason
 
-    // get ff
-    _vel_ff = _cmd_vel*_param_fwd_k_ff;
+    // Write to data flash log
+    AP::logger().WriteStreaming("AROT",
+                                "TimeUS,LR",
+                                "s-",
+                                "F-",
+                                "QB",
+                                AP_HAL::micros64(),
+                                _landed_reason);
+}
+#endif  // HAL_LOGGING_ENABLED
 
-    //calculate acceleration target based on PI controller
-    _accel_target = _vel_ff + _vel_p;
+// Arming checks for autorotation, mostly checking for miss-configurations
+bool AC_Autorotation::arming_checks(size_t buflen, char *buffer) const
+{
+    if (!enabled()) {
+        // Don't run arming checks if not enabled
+        return true;
+    }
+
+    const AP_HAL::HAL& hal = AP_HAL::get_HAL();
```

Comment:
Typically we just do this at the top of the file with `extern const AP_HAL::HAL& hal;`

Reply:
Done